File: NationalPokedex/NationalPokedex/Models/PokeApiModel.swift
//
//  PokeApiModel.swift
//  NationalPokedex
//
//  Created by Akihiro Matsuyama on 2024/04/05.
//

import Foundation

protocol PokeApiModelProtocol {
    /// 指定されたポケモンのリストを更新し、新しいポケモンのデータを非同期に取得する
    /// - Parameter pokemons: 現在のポケモンの配列
    /// - Returns: 更新後のポケモンの配列を含むResultオブジェクト
    func getNewPokemons(pokemons: [Pokemon]) async -> (Result<[Pokemon], ApiError>)
}

struct PokeApiModel: PokeApiModelProtocol {
    enum Endpoint: String {
        case pokemon = "pokemon"
        case pokemonSpecies = "pokemon-species"

        static let baseURL = "https://pokeapi.co/api/v2/"

        func url(for id: String) -> URL? {
            URL(string: "\(Endpoint.baseURL)\(self.rawValue)/\(id)")
        }
    }

    func getNewPokemons(pokemons: [Pokemon]) async -> Result<[Pokemon], ApiError> {
        // ポケモン配列が0の場合は図鑑番号1番から、内容が存在する場合は最後のポケモン番号+1から
        let startPokemonId = pokemons.isEmpty ? 1 : pokemons.last!.id + 1
        let newPokemonsResult = await getMultiplePokemon(startId: startPokemonId, count: 20)

        switch newPokemonsResult {
        case .success(let newPokemons):
            return .success(pokemons + newPokemons)
        case .failure(let error):
            return .failure(error)
        }
    }

    private func getMultiplePokemon(startId: Int, count: Int) async -> Result<[Pokemon], ApiError> {
        var pokemonList: [Pokemon] = []
        var encounteredError: ApiError?

        for id in startId..<(startId + count) {
            let result = await getNewPokemon(id: id.description)
            switch result {
            case .success(let pokemon):
                pokemonList.append(pokemon)
            case .failure(let error):
                print("Failed to get Pokemon with id \(id): \(error)")
                encounteredError = error
            }
        }

        if let error = encounteredError {
            return .failure(error)
        } else {
            return .success(pokemonList)
        }
    }


    private func getNewPokemon(id: String) async -> (Result<Pokemon, ApiError>) {
        let individualResult: Result<PokeIndividual, ApiError> = await self.getPokeData(id: id, endpoint: .pokemon)
        let speciesResult: Result<PokeSpecies, ApiError> = await self.getPokeData(id: id, endpoint: .pokemonSpecies)

        switch (individualResult, speciesResult) {
        case (.success(let individualResult), .success(let speciesResult)):
            return .success(Pokemon(id: individualResult.id, individual: individualResult, species: speciesResult))

        case (.failure(let error), _), (_, .failure(let error)):
            return .failure(error)
        }
    }

    private func getPokeData<T: Decodable>(id: String, endpoint: Endpoint) async -> Result<T, ApiError> {
        guard let url = endpoint.url(for: id) else {
            return .failure(.urlError)
        }

        do {
            let (data, _) = try await URLSession.shared.data(for: URLRequest(url: url))
            do {
                let decodedData = try JSONDecoder().decode(T.self, from: data)
                return .success(decodedData)
            } catch {
                return .failure(.jsonParseError(String(data: data, encoding: .utf8) ?? "Invalid data"))
            }
        } catch {
            return .failure(.responseError(error))
        }
    }
}

File: NationalPokedex/shared/src/commonMain/kotlin/com/akidon0000/nationalpokedex/Api/PokeApiModel.kt
package com.akidon0000.nationalpokedex.Api

import com.akidon0000.nationalpokedex.Model.*
import io.ktor.client.HttpClient
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.serialization.kotlinx.json.json
import io.ktor.client.request.*
import io.ktor.client.call.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer
import kotlin.reflect.KClass

interface PokeApiModelProtocol {
    suspend fun getNewPokemons(pokemons: List<Pokemon>): Result<List<Pokemon>, ApiError>
}

class PokeApiModel : PokeApiModelProtocol {
    enum class Endpoint(val path: String) {
        POKEMON("pokemon"),
        POKEMON_SPECIES("pokemon-species");

        companion object {
            const val BASE_URL = "https://pokeapi.co/api/v2/"
        }

        fun url(id: String): String {
            return "$BASE_URL${this.path}/$id"
        }
    }

    override suspend fun getNewPokemons(pokemons: List<Pokemon>): Result<List<Pokemon>, ApiError> {
        val startPokemonId = if (pokemons.isEmpty()) 1 else pokemons.last().id + 1
        val newPokemons = getMultiplePokemon(startId = startPokemonId, count = 20)
        return Result.success(newPokemons)
    }

    private suspend fun getMultiplePokemon(startId: Int, count: Int): List<Pokemon> {
        val pokemonList = mutableListOf<Pokemon>()

        for (id in startId until (startId + count)) {
            val result = getNewPokemon(id.toString())
            when (result) {
                is Result.success -> pokemonList.add(result.value)
                is Result.failure -> println("Failed to get Pokemon with id $id: ${result.error}")
            }
        }
        return pokemonList
    }

    private suspend fun getNewPokemon(id: String): Result<Pokemon, ApiError> {
        val individualResult: Result<PokeIndividual, ApiError> = getPokeData(id, Endpoint.POKEMON, PokeIndividual::class)
        val speciesResult: Result<PokeSpecies, ApiError> = getPokeData(id, Endpoint.POKEMON_SPECIES, PokeSpecies::class)

        return when {
            individualResult is Result.success && speciesResult is Result.success -> {
                Result.success(Pokemon(id = individualResult.value.id, individual = individualResult.value, species = speciesResult.value))
            }
            individualResult is Result.failure -> Result.failure(individualResult.error)
            speciesResult is Result.failure -> Result.failure(speciesResult.error)
            else -> Result.failure(ApiError.UrlError)
        }
    }

    @OptIn(kotlinx.serialization.InternalSerializationApi::class)
    private suspend fun <T> getPokeData(id: String, endpoint: Endpoint, type: KClass<T>): Result<T, ApiError> where T : Any {
        val url: String = endpoint.url(id)

        return try {
            val json = Json { ignoreUnknownKeys = true }
            val response: String = httpClient.get(url).body()
            val pokemon: T = json.decodeFromString(type.serializer() , response)
            Result.success(pokemon)
        } catch (e: Exception) {
            println("ERROR: $e")
            Result.failure(ApiError.JsonParseError(e.toString()))
        }
    }

    private val httpClient = HttpClient {
        install(ContentNegotiation) {
            json(Json {
                prettyPrint = true
                isLenient = true
                ignoreUnknownKeys = true
            })
        }
    }
}

